### PR TITLE
[CALCITE-4199] review-only: nullability annotations

### PR DESCRIPTION
**Pull request by vlsi**
This PR splits https://github.com/apache/calcite/pull/2268 into "Refactor: prepare code for nullability annotations" and "Add @nullable annotations only
dc6ef83" commits.

It might be slightly easier for the review, however, the key commit is still big https://github.com/apache/calcite/pull/2273/commits/2e00e9cf7fd7e2768ae11e383b1cc509306c5067

```
757 files changed, 7487 insertions(+), 3827 deletions(-)
```

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/adapter/clone/ArrayTable.java [hunk 1]
```diff
@@ -243,9 +242,9 @@ public static List asList(final Representation representation,
 
     /** Converts a value set into a compact representation. If
      * {@code sources} is not null, permutes. */
-    Object freeze(ColumnLoader.ValueSet valueSet, int[] sources);
+    Object freeze(ColumnLoader.ValueSet valueSet, int @Nullable [] sources);
```
Comment (liyafan82) [inline core/src/main/java/org/apache/calcite/adapter/clone/ArrayTable.java:245]: It seems some annotations appear before the type, while others appear after the type. We should make this consistent?

File: core/src/main/java/org/apache/calcite/adapter/clone/ArrayTable.java [hunk 2]
```diff
@@ -446,25 +451,30 @@ public static List asList(final Representation representation,
       return Pair.of(codes, codeValues);
     }
 
+    private static Pair<Object, @Nullable Comparable[]> unfreeze(Object value) {
```
Comment (liyafan82) [inline core/src/main/java/org/apache/calcite/adapter/clone/ArrayTable.java:454]: Maybe it is helpful to replace Object with a generic type?

File: core/src/main/java/org/apache/calcite/adapter/clone/ArrayTable.java (hunk 1 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/adapter/clone/ArrayTable.java:245]: Please read https://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html#jls-9.7.4 amd 
https://github.com/vlsi/calcite/blob/checkerframework/site/develop/index.md#null-safety

`@Nullable int[]` and `int @Nullable []` are different.

File: core/src/main/java/org/apache/calcite/adapter/clone/ArrayTable.java (hunk 2 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/adapter/clone/ArrayTable.java:454]: The point of the PR is nullability annotations, not code refactoring. In my opinion, `Pair` should be replaced with properly named classes.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java [hunk 3]
```diff
@@ -1049,19 +1063,19 @@ private boolean isOverlapped(Pair<Long, Long> a, Pair<Long, Long> b) {
     @Override public void close() {
     }
 
-    private Object[] takeOne() {
-      return list.pollFirst();
+    private @Nullable Object[] takeOne() {
+      return requireNonNull(list.pollFirst(), "list.pollFirst()");
```
Comment (liyafan82) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java:1067]: This seems contradictive? `requireNonNull` indicates it should be non-nullable?

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java:1067]: This is just fine. `pollFirst` can return `null`, and here `non-null` is required by `takeOne`

**Pull request closed without merging**